### PR TITLE
Useful listing table

### PR DESCRIPTION
**Pull request by jelly**
image

* [x] Make selectable visually work `.selected-class-thing + td { background: var(--pf-v5--theBackGroundColor); }
`
* [x] Stop using a Card in the first column, `<Icon isInline><p>File or dir</p>` instead
* [x] Make the contextmenu and selecting event handlers work
* [x] Make everything sortable
* [x] Stretch goal, combine both views into one as per https://garrett.github.io/cockpit-mockweb/navigator/grid#

## Re-useable table approach issues

* [ ] Refactor / clean up render function
* [x] Size header is not allowed with content
image
* [x] Directory is empty twice
image
* [x] Too much spacing when there three items
image
* [x] wrong icon for folders
* [x] Icon view not really centered
image
* [x] Icon not centered / wrongly aligned
image
* [x] General text/font

## Follow up

* [ ] More useful dates? "Yesterday at 12:00" instead?
* [ ] Update & Merge the PF update PR as we no longer use a <Card> which blocked us from updating
* [ ] Scrollbar position wrong
image
* [ ] Table should take the full height when available.
image
`.pf-v5-c-sidebar__main, .pf-v5-c-sidebar__content, .pf-v5-c-card { height: 100%; }`

### CODE REVIEW

**Comment by jelly** [comment]
Selected state needs some CSS love:

image

**Comment by jelly** [comment]
@garrett I have come quite far porting your CSS to the file viewer one big interesting detail is that our tables are a bit weird:
```
<tbody>
<tr>
</tr>
</tbody>
<tbody>
...
````

**Review by allisonkarlitskaya — COMMENTED** [review]
Some assorted drive-bys:

File: src/files-card-body.jsx [hunk 1]
```diff
@@ -342,6 +355,27 @@ export const FilesCardBody = ({
...
+            direction: filterColumnMapping[sortBy][1],
+        },
+        onSort: (_event, index, direction) => {
+            setSortBy(filterColumns[index][direction].itemId);
+        },
+        columnIndex,
+    });
+
+    const getFileType = (file) => {
+        if (file.type === "dir") {
```
[the remark this answers is not in the document]
Comment (jelly) [reply src/files-card-body.jsx:370]: Hmmm if that is the case we should refactor some more code. Let's also see if we have a test for this.

File: src/files-card-body.scss [hunk 2]
```diff
@@ -0,0 +1,179 @@
...
+            background-color: var(--color-icon);
+            block-size: var(--icon-size);
+            content: '';
+            mask-repeat: no-repeat;
+            mask-position: center;
+            mask-image: url('data:image/svg+xml,<svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 1536 1792"><path d="M1468 380c37 37 68 111 68 164v1152c0 53-43 96-96 96H96c-53 0-96-43-96-96V96C0 43 43 0 96 0h896c53 0 127 31 164 68zm-444-244v376h376c-6-17-15-34-22-41l-313-313c-7-7-24-16-41-22zm384 1528V640H992c-53 0-96-43-96-96V128H128v1536z"/></svg>');
+        }
+    }
+
+    tr[data-type="folder"] {
```
[the remark this answers is not in the document]
Comment (jelly) [reply src/files-card-body.scss:54]: Symlinks should get an extra class as the idea is to overlay a small svg. Let's already add that.

Icons hmm we could indeed set a specific data-type that doesn't sound too bad.

File: src/files-card-body.jsx [hunk 3]
```diff
@@ -73,6 +74,14 @@ const compare = (sortBy) => {
                 ? -1
                 : 1)
             : compareFileType(a, b);
+    case "largest_size":
+        return (a, b) => (a.size > b.size
+            ? -1
```
[the remark this answers is not in the document]
Comment (jelly) [reply src/files-card-body.jsx:79]: It does not, and in my opinion why would directories need to be first if I am sorting on size. We have no file count information or directory size information so I don't see how sorting on file size is useful in that regard by listing directories first. (@garrett )

[the remark this answers is not in the document]
Comment (jelly) [reply src/files-card-body.jsx:79]: Not going to block ton this but it doesn't feel like it makes sense to me.

[the remark this answers is not in the document]
Comment (jelly) [reply src/files-card-body.jsx:79]: If you mean it doesn't handle two items equal and then sort them alphabetically? Then yes that is true :)

File: src/files-card-body.jsx [hunk 4]
```diff
@@ -342,94 +351,95 @@ export const FilesCardBody = ({
         </ContextMenu>
     );
 
+    const sortColumn = (columnIndex) => ({
+        sortBy: {
+            index: filterColumnMapping[sortBy][0],
+            direction: filterColumnMapping[sortBy][1],
+        },
+        onSort: (_event, index, direction) => {
+            setSortBy(filterColumns[index][direction].itemId);
```
Comment (cockpituous) [inline src/files-card-body.jsx:360]: These 2 added lines are not executed by any test.

Comment (cockpituous) [inline src/files-card-body.jsx:360]: These 2 added lines are not executed by any test.

File: src/files-card-body.jsx [hunk 5]
```diff
@@ -342,94 +351,95 @@ export const FilesCardBody = ({
...
-    function getFileType(file) {
-        if (file.type === "dir") {
-            return "directory-item";
-        } else if (file.type === "lnk" && file?.to === "dir") {
-            return "directory-item";
-        } else {
-            return "file-item";
-        }
+const getFileType = (file) => {
+    if (file.type === "dir" || file.to === "dir") {
```
Comment (allisonkarlitskaya) [inline src/files-card-body.jsx:403]: Did a simple `file.to === 'dir'` not work?

File: src/files-card-body.jsx [hunk 6]
```diff
@@ -342,94 +351,95 @@ export const FilesCardBody = ({
...
+        return "folder";
+    } else {
+        return "file";
     }
+};
+
+// Memoize the Item component as rendering thousands of them on each render of parent component is costly.
+const Row = React.memo(function Item({ file, isSelected }) {
```
Comment (allisonkarlitskaya) [inline src/files-card-body.jsx:411]: For the record, I don't think the memoize currently protects us from changes to unrelated files since we re-create the list on every single fsinfo event, but I'm happy to see what we are theoretically memoized, because it means that we can fix that bug and benefit from it...

File: src/files-card-body.scss [hunk 7]
```diff
@@ -0,0 +1,195 @@
...
+                + td {
+                    padding-block-end: var(--pf-v5-global--spacer--sm);
+                }
+            }
+
+            .item-num {
+                display: none;
+            }
+
+            .item-size,
```
Comment (allisonkarlitskaya) [inline src/files-card-body.scss:179]: Cool!!!

File: src/files-card-body.scss [hunk 8]
```diff
@@ -0,0 +1,195 @@
...
+            .item-date {
```
Comment (allisonkarlitskaya) [inline src/files-card-body.scss:180]: I think the inclusion of `.item-date` here is probably accidental since you `display: none` it, just below?

**Review by allisonkarlitskaya — APPROVED** [review]
This looks great!

Some minor comments below, but feel free to handle them as follow-ups...

File: src/files-card-body.jsx (hunk 5 above)
Comment (jelly) [reply src/files-card-body.jsx:403]: So you mean only `file.to`?

Comment (allisonkarlitskaya) [reply src/files-card-body.jsx:403]: Yes, like the `file.type === "dir"` part here is redundant.  If `file.type === "dir"` then `file.to` will also `=== "dir"`.

File: src/files-card-body.jsx (hunk 6 above)
Comment (jelly) [reply src/files-card-body.jsx:411]: It does, I tested this on /bin, the event listener took ~ 200-300ms and with memozing its a lot faster.

React can be a bit confusing :)

So re-render => props/state changes and the render function is called again forcing all children to "re-render". By memozing React does not re-run the `render` function as the props did not change.

File: src/files-card-body.scss (hunk 8 above)
Comment (jelly) [reply src/files-card-body.scss:180]: Howso, we want the font-size and color to be the same?

Comment (allisonkarlitskaya) [reply src/files-card-body.scss:180]: But we don't show the date...

File: src/files-card-body.jsx (hunk 4 above)
Comment (cockpituous) [inline src/files-card-body.jsx:360]: These 2 added lines are not executed by any test.

Comment (cockpituous) [inline src/files-card-body.jsx:360]: These 2 added lines are not executed by any test.

**Review by allisonkarlitskaya — APPROVED** [review]
As a wise man once said...

**Comment by allisonkarlitskaya** [comment]
yolo.

File: src/files-card-body.jsx (hunk 6 above)
Comment (jelly) [reply src/files-card-body.jsx:411]: Another issue might be that we memoize the sort order of the array. But then again changing our state to an Object might help with that :)
